### DISCUSSION TITLE
Query user role but only get the root roles

### DISCUSSION THREAD

**Discussion by falconSDF** [casdoor/casdoor#2974] [Q&A] [answered]
I use role in an inheritive way，literally speaking，I have Basic_role， Manager_role, Supervisor_role, Root_role. I set Manager_role and  Supervisor_role as sub roles of Basic_role, and set Basic_role as sub role of Manager_role and Supervisor_role.

My purpose is that I can assign basic_permission to Basic_role, and by setting sub roles Manager_role and Supervisor_role will automatically get basic_permission, and the same for the two roles and Root_role. 
I assign Root_role to  root_user, and it works well when I call enforcer, root_user can access all resources under  basic_permission, manager_permission and supervision_permission.

However, when I try to fetch user information ,  it went wrong.  When I vistit user info page (related api is like  "/api/get-user?id=built-in/jackblack369"), it only returns Root_role and Basic_role. I tried many times, finding that it only returns the role that is directly bounded to the user, and the top-level roles founded through sub-rule-relationship.

I read the code in the  folloing trace:

routers.Router.go:           beego.Router("/api/get-user", &controllers.ApiController{}, "GET:GetUser")
controllers.User.go:        object.ExtendUserWithRolesAndPermissions(user)
object.User.go:               user.Permissions, user.Roles, err = getPermissionsAndRolesByUser(user.GetId())
object.Permission.go:           roles, err := getRolesByUser(userId)
object.User.go:                ,   allRoles, err := GetAncestorRoles(allRolesIds...)
At the end,  the roles are returned by the GetAncestorRoles. 

I read through the code, making sure that it will return all ancestor roles of Root_role, which conflits with the behavior.

The casdoor I run is   docker image        casbin/casdoor :latest with image ID fadd8f0b14ef.

Do you guys behave the same in this case?  Did I miss anything in the code or some updates has been made after my image version?

![image](https://github.com/casdoor/casdoor/assets/51161126/563103de-7d75-4457-b10d-8631def9063f)

![image](https://github.com/casdoor/casdoor/assets/51161126/a0119874-f25f-48ed-ab8d-735cc5b17bbb)

**Comment by hsluoyz** [comment, marked as answer]
@falconSDF use Casbin API `GetImplicitRolesForUser()`: https://casbin.org/docs/rbac-api#getimplicitrolesforuser
